### ISSUE TITLE
When SFDIR is in use, saving font shouldn't overwrite every glyph

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#3779]
This ties into #3778. The bigger the font, due partly to #3778 but also perhaps due to having many images in the SFD from manuscripts, books, other primary sources, the longer it takes to save.

It seems like saving as an SFDIR is a way to avoid this problem. However, it is not.

FontForge already has a robust system of knowing when a font has had a glyph changed or not. But this system isn't used when you save your SFDIR, every glyph is overwritten every time, changed or not; SFDIR is just as slow as SFD!

I think that only the changed glyphs should be saved, and `font.props` should be saved every time.

# Technical info

The function most relevant to this request is `SFDWrite`. Right now it always calls `SFDirClean` if saving to a directory; that function deletes all the files it "recognize" (all the glyphs). Seems like the loop beginning @sfd.c:2876 is the most obvious thing to change, along with not deleting any files.

**Comment by khaledhosny** [comment]
Deleting glyph files is so that when a glyph is deleted in the font it gets deleted on disk, otherwise next time the font is loaded the deleted glyph will be read again by FontForge and potentially conflict with other glyphs in the font which would mess things up.

There are probably a less obtrusive ways to handle this.

**Comment by ctrlcctrlv** [comment]
@khaledhosny Right&hellip;one way to handle that would be to delete all the glyph files (files ending in `.glyph`) not in the current font. That would mean we don't have to overwrite every file. What do you think @khaledhosny ?

**Comment by skef** [comment]
It's not very hard to:

1. Generate and sort a list of files that would be saved to a new location.
2. Generate and sort a list of files that exist in the directory.
3. Delete any file in the second list and not in the first list. 

The main decision is whether to do that for every file in the relevant directory, or only those files with a certain extension. I generally favor the extension approach, because then users can temporarily rename files they want to get out of the way and the program doesn't go around deleting them.

**Comment by skef** [comment]
Actually, hell, easier to hash than to sort if you have an implementation handy.

**Comment by ctrlcctrlv** [comment]
FontForge probably has a hash function somewhere in its codebase, considering GW's method of programming, hehe.

**Comment by frank-trampe** [comment]
@ctrlcctrlv, that's true.

Do you actually use sfdir? Does it actually work well? It has gone unmaintained for a long time.

**Comment by ctrlcctrlv** [comment]
I started using it with my Quaerite Regnum Dei font because I thought it would fix my problem (as written). Other then the issue at hand it seems to work well.

**Comment by ctrlcctrlv** [comment]
Thinking about this issue more (i.e. before writing any code, but after reading the relevant code) I also think we should overwrite every glyph if the font's encoding changed since the last time we wrote out the font, because that would change the glyph's `Encoding:` line even if it hasn't changed.

By the way @frank-trampe , I know you wrote the UFO code, so you would prefer people just use that, but UFO (or at least FontForge's implementation of it) doesn't support Spiro, so all my Spiro glyphs lose their data, so I can't use that. It also doesn't support User Decompositions, but that might be my fault (I didn't see any relevant part of the UFO spec when implementing them).

**Comment by frank-trampe** [comment]
I realize that UFO is not great as a working format. (It also omits a lot of data, as you note.) I just see the upsides of sfdir over sfd as extremely limited absent a lot of improvements to realize the advantages of the separate file structure, particularly since sfd diffs so well.

There was a request a while back to implement selective writing of UFO like this (and then to allow true saving to UFO). I don't remember a lot of the details, but I recall that there were a lot of wrinkles and that using the FontForge glyph changed flag was not a sufficient marker of whether to rewrite the glif file. I'll try to find my notes tomorrow.

**Comment by ctrlcctrlv** [comment]
Can anyone think of another reason that we'd need to write out the glyph data even if the glyph hasn't been marked by FontForge as changed? Perhaps OpenType table changes? (It might make more sense to update FontForge to mark glyphs changed in these situations rather than write special cases for them.)

**Comment by ctrlcctrlv** [comment]
Thanks @frank-trampe , commented on an old page so didn't see yours.

**Comment by ctrlcctrlv** [comment]
By the way @frank-trampe , if I write a patch for this, I'm going to recommend we release this into the wild in phases:

1. Add Option `FastSFDIRSave`, off by default. Option description: &laquo;Save your SFDIR files faster by not overwriting glyphs which haven't changed. This feature should be considered alpha-quality and may ruin your files, keep good backups, thanks for helping us test it.&raquo;
2. After a while of no reported issues, make `FastSFDIRSave` on by default. Warn users who last saved their SFDIR with an old version of FontForge about this change. Probably we can do this just by bumping the SFD version.
3. If still no reported issues, remove the option, assume SFDIR writing glyph by glyph is safe, any bugs in it will just be treated as bugs in SFDIR generally.

**Comment by skef** [comment]
> Can anyone think of another reason that we'd need to write out the glyph data even if the glyph hasn't been marked by FontForge as changed?

The general reason is that it's easy to miss something, and always writing out everything fixes that problem. 

A good general way to handle this problem is to have a normalized form of the data and a "checksum of checksums" on that form. For smaller piles of data (almost everything) you normalize and checksum it. For large piles (like images) you keep its checksum around with it (both in memory and when storing) and calculate on changes. Then you use a checksum of checksums as the master record of what's in the file.

**Comment by ctrlcctrlv** [comment]
No, I get that. I'm saying if/when I change it, what other cases should I be aware of when writing the patch. I understand that the current way makes things easier. Unfortunately my SFD has too many images to do things the easy way

**Comment by khaledhosny** [comment]
> Do you actually use sfdir? Does it actually work well? It has gone unmaintained for a long time.

I do and have been doing for years, no issues. I don't think there is much that needs to be maintained about it since it is basically SFD witch each glyph in its own file.

**Comment by khaledhosny** [comment]
Another issue when saving only changed file is that if unchaged glyph was modified on disk by another process, FontForge would not know about this and issue won't show until the next time the file is loaded, with potential data loss.

**Comment by khaledhosny** [comment]
The current FontForge behavior, AFAICT, is that save always saves everything and does not try to check whether the font was changed or not (both when saving SFD and SFDIR), so the proposed solution would be changing this for SFDIR only which is rather inconsistent and might be unexpected by users.

**Comment by ctrlcctrlv** [comment]
If the change is controversial, I'd be willing to accept adding an option and leaving it false by default into perpetuity.

**Comment by ctrlcctrlv** [comment]
However @khaledhosny you should be able to understand why this change is desirable. If you don't take a look at this screenshot. This is how I'm using FontForge.

![2019-07-12-182705_3840x2160_scrot](https://user-images.githubusercontent.com/838783/61121737-8d2af180-a4d2-11e9-831f-0bc2aa2a6c26.png)

Am I using it wrong? I don't think so&hellip;George Williams also made fonts based on manuscripts, and as all my fonts except one are _restorations_, they are all based on images, though some of my older work was made in Illustrator.

You get enough of these glyphs in the font and eventually you have a 5GB SFDIR.

**Comment by ctrlcctrlv** [comment]
Nevermind, this is way too difficult, I'm closing this in favor of #3778. #3778 would make it so the overwrite performance wouldn't matter.

**Comment by khaledhosny** [comment]
I wasn't saying the change is bad, I pointed out to potential issues to watch out for. I'm right now running Ubuntu on WSL and the file access is abysmal, saving an SFDIR with 6000+ glyphs takes ages and I'd be very welcome to any improvements in this area.

**Comment by ctrlcctrlv** [comment]
Is your issue due to images? If not I'll reopen this.

**Comment by khaledhosny** [comment]
No images just lots of glyphs (it was not an issue with Linux and SSD, so it is really a IO issue).
